CMakeLists.txt

```diff
@@ -0,0 +1,114 @@
+    # Set custom compiler flags for gcc-compatible compilers
+    set(CMAKE_CXX_FLAGS_RELEASE "-O3 -DNDEBUG -funroll-loops -fcx-limited-range")
+    set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g -DNDEBUG -funroll-loops -fcx-limited-range")
+endif()
+
+include(CTest)
+
+set(FINUFFT_ARCH_FLAGS "-march=native" CACHE STRING "Compiler flags for specifying target architecture.")
+set(FINUFFT_FFTW_SUFFIX "OpenMP" CACHE STRING "Suffix for FFTW libraries (e.g. OpenMP, Threads etc.)")
+set(FINUFFT_BUILD_EXAMPLES "OFF" CACHE BOOL "Whether to build the FINUFFT examples")
+set(FINUFFT_BUILD_TESTS "OFF" CACHE BOOL "Whether to build the FINUFFT tests")
+set(FINUFFT_BUILD_FORTRAN "OFF" CACHE BOOL "Whether to build the FINUFFT Fortran examples")
+set(FINUFFT_BUILD_MATLAB "OFF" CACHE BOOL "Whether to build the FINUFFT Matlab interface")
+set(FINUFFT_ENABLE_SANITIZERS "ON" CACHE BOOL "Whether to enable sanitizers, only effective for Debug configuration.")
+set(FINUFFT_USE_OPENMP "ON" CACHE BOOL "Whether to use OpenMP for parallelization. If disabled, the finufft library will be single threaded. This does not affect the choice of FFTW library.")
+
+if(FINUFFT_BUILD_MATLAB)
+    # When building for matlab, we will fetch the OpenMP library used by matlab
+    # instead of system default for compatibility.
+    find_package(Matlab REQUIRED)
+    find_library(OpenMP_iomp5_LIBRARY NAMES iomp5 HINTS ${Matlab_ROOT_DIR}/sys/os/ PATH_SUFFIXES glnxa64 maci64)
+    find_library(OpenMP_pthreads_LIBRARY NAMES pthread CMAKE_FIND_ROOT_PATH_BOTH)
+endif()
+
+# Find required packages
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+find_package(FFTW REQUIRED)
+
+if(FINUFFT_USE_OPENMP)
+    find_package(OpenMP REQUIRED)
+endif()
+
+set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+
+# This set of sources is compiled twice, once in single precision and once in double precision
+# The single precision compilation is done with -DSINGLE
+set(FINUFFT_PRECISION_DEPENDENT_SOURCES src/finufft.cpp src/simpleinterfaces.cpp src/spreadinterp.cpp src/utils.cpp fortran/finufftfort.cpp)
+
+# Set of compilers which behave like gcc
+set(FINUFFT_GNU_LIKE_COMPILERS AppleClang Clang GNU)
+
+# Utility function to enable ASAN on debug builds
+function(enable_asan target)
+    if(NOT FINUFFT_ENABLE_SANITIZERS)
+        return()
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID IN_LIST FINUFFT_GNU_LIKE_COMPILERS)
+        # Enable only on clang / gcc compilers.
+        target_compile_options(${target} PRIVATE $<$<CONFIG:DEBUG>:-fsanitize=address>)
+        target_link_options(${target} PRIVATE $<$<CONFIG:DEBUG>:-fsanitize=address>)
+    endif()
+endfunction()
+
+# Utility function to set finufft compilation options.
+function(set_finufft_options target)
+    set_property(TARGET ${target} PROPERTY POSITION_INDEPENDENT_CODE ON)
+    set_property(TARGET ${target} PROPERTY CMAKE_CXX_STANDARD 14)
```

Comment:
For perfromance reasons we could define `__STDCPP_DEFAULT_NEW_ALIGNMENT__` here if AVX is supported for example:
```cmake
if ("cxx_std_20" IN_LIST CMAKE_CXX_COMPILE_FEATURES)
    message(STATUS "Using C++20 standard")
    set(CMAKE_CXX_STANDARD 20)
elseif ("cxx_std_17" IN_LIST CMAKE_CXX_COMPILE_FEATURES)
    set(CMAKE_CXX_STANDARD 17)
    message(STATUS "Detected support for C++17 standard")
elseif ("cxx_std_14" IN_LIST CMAKE_CXX_COMPILE_FEATURES)
    set(CMAKE_CXX_STANDARD 14)
    message(STATUS "Detected support for C++14 standard")
else ()
    message(FATAL_ERROR "Cannot detect CXX_STANDARD of C++14 or newer.")
endif ()

if (CMAKE_CXX_STANDARD GREATER 14)
  if (ENABLE_AVX2)
    foreach (FLAG "-mavx2" "/arch:AVX2")
      unset (HAVE_AVX2 CACHE)
      check_c_compiler_flag (${FLAG} HAVE_AVX2)
      if (HAVE_AVX2)
        set (AVX2_FLAG ${FLAG})
        break()
      endif ()
    endforeach ()
    if(AVX2_FLAG)
        add_definitions(__STDCPP_DEFAULT_NEW_ALIGNMENT__=32)
    endif()
  endif ()
  if (NOT AVX2_FLAG)
    foreach (FLAG "-mavx" "/arch:AVX")
        unset (HAVE_AVX CACHE)
        check_c_compiler_flag (${FLAG} HAVE_AVX)
        if (HAVE_AVX)
        set (AVX_FLAG ${FLAG})
        break()
        endif ()
    endforeach ()
    if(AVX_FLAG)
        add_definitions(__STDCPP_DEFAULT_NEW_ALIGNMENT__=16)
    endif()
  endif()
endif()
```

Reply:
I do not believe that redefining `__STDCPP_DEFAULT_NEW_ALIGNMENT__` does what you expect it does. According to C++ standard 19.8.1, it is a pre-defined macro to represent: "An integer literal of type `std::size_t` whose value is the alignment guaranteed by a call to `operator
new(std::size_t)` or `operator new[](std::size_t)`.

The user is not allowed to redefine its value according to 19.8.4:
"If any of the pre-defined macro names in this subclause, or the identifier defined, is the subject of a `#define`
or a `#undef` preprocessing directive, the behavior is undefined"

In any case, this PR is a systems PR in preparation for us to rework performance-critical parts of the projects with explicit hand-written vectorization, during which we will also take a look at all other performance related aspects (alignment etc.)